ldms/src/ldmsd/ldmsd_request.c

```diff
@@ -766,6 +767,29 @@ static int string2attr_list(char *str, struct attr_value_list **__av_list,
 	return rc;
 }
 
+int validate_ldmsd_req(ldmsd_req_ctxt_t reqc, ldmsd_req_hdr_t rh)
+{
+	ldmsd_req_attr_t attr;
+	unsigned int off;
+	/* Validate length of request header and request. */
+	if (rh->rec_len < sizeof(*rh))
+		return 0;
+	attr = (ldmsd_req_attr_t)(rh+1);
+	off = sizeof(*rh);
+	while (attr->discrim)
+	{
+		off += sizeof(*attr) + attr->attr_len;
+		/* Check attr doesn't reference beyond end of request buffer */
+		if (off > rh->rec_len)
+			return 0;
+		/* Check null termination of each attr_value */
+		else if (attr->attr_value[attr->attr_len-1] != '\0')
+			return 0;
+		attr = (ldmsd_req_attr_t)&attr->attr_value[attr->attr_len];
```

Comment:
And here again we need to make sure that this new "attr" fits in the rh->rec_len before we go accessing attr->discrim in the while() statement. It may make sense to rework this as a do-while or for loop so you don't need to put that check in two different places.

Reply:
You're right, I think after line 788
```
if (attr->discrim && off + sizeof(*attr) > rh->rec_len)
    return 0;
```
At that point attr_len can be safely checked next time around the loop. And yes, perhaps refactoring the logic as a do-while could make things simpler.